### PR TITLE
Fix timer drift under high UDP traffic conditions

### PR DESCRIPTION
**Pull request by jjneely**
The initial problem here is that in high UDP traffic situations the
event loop would spend multiple seconds, even 30 or 45+, reading UDP
data in handle_udp_message().  As the event loop didn't make iterations
this would push out when timer events ran.  This would cause gaps in
Graphite data immediately followed by an artificial spike where summary
metrics were generated for a window much longer than the configured
flush interval.

This tunes up handle_udp_message() to make use of the 64 KiB circular
buffer.  The function reads until the buffer is full (assuming max
packet size is 1500 bytes) or the read would block.  I've removed any
memory allocation from the loop as well.  This is level triggered so
the next poll cycle will continue processing UDP data if ready.  This
insures that we make progress through all events.

The benchmarking I've done shows an improvement from about 350K/400K
packets per second with large amounts of packet drop before this patch
to much lower packet drop and 550K packets per second are processed.
Timer drift has not been observed.

I've patched handle_binary_client_connect() to handle '\n' chars as
inserted by the UDP handler.

Potential bugs:  This doesn't handle cases where the binary protocol and
ascii protocol are interleaved in a single buffer.

### CODE REVIEW

File: src/networking.c [hunk 1]
```diff
@@ -438,18 +438,22 @@ static int read_client_data(conn_info *conn) {
...
 
-        // Clear the input buffer
-        circbuf_clear(&conn->input);
+    // Clear the input buffer
+    circbuf_clear(&conn->input);
 
+    // Loop until our buffer cannot hold another UDP packet (using a 1500
+    // byte MTU) or we cannot read another packet off the wire.  We do not
+    // read continuously off of the UDP socket to preserve timer execution.
+    while (circbuf_avail_buf(&conn->input) > 1500) {
```
Comment (johnkeates) [inline src/networking.c]: What would happen if for some reason the MTU would be different? Not sure if anyone is going to enable Jumbo frames on an Ethernet interface for Statsite, or maybe puts Statsite on a PPPoE connection... but you know, hardcoded assumptions and stuff. I myself think it's fine and putting a note in the docs would be fine as well, anyone else have anything to say about it?

**Comment by johnkeates** [comment]
By the way, in general, I think this patch is fine, and we're only imposing 2 possible bugs. Releasing it as 0.9.0 with my upcoming patch for init/systemd scripts and user setup might be fine in case this is breaking for anyone.

File: src/networking.c (hunk 1 above)
Comment (jjneely) [inline src/networking.c]: I had similar thoughts, but wasn't sure where to find/build a better value.  UDP packet size is a fairly well understood StatsD limitation, so I'm leaning toward just documenting it.

Comment (johnkeates) [inline src/networking.c]: Docs it is. I'll wait a few hours to see if someone else has something to say, and then we'll just merge it in.

Comment (jjneely) [inline src/networking.c]: See PR #234 for doc updates.

**Pull request merged**